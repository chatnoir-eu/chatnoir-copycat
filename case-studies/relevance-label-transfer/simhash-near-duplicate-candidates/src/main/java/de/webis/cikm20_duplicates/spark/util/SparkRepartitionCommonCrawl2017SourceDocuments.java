package de.webis.cikm20_duplicates.spark.util;

import org.apache.hadoop.io.compress.BZip2Codec;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import de.webis.cikm20_duplicates.spark.SparkCreateSourceDocuments;

public class SparkRepartitionCommonCrawl2017SourceDocuments {
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			String corpus = "cc-2017-04";
			
			for(String c: SparkCreateSourceDocuments.cc17Collections()) {
				context.textFile("cikm2020/document-fingerprints-final/" + c + "-jsonl")
					.repartition(50)
					.saveAsTextFile("cikm2020/document-fingerprints-final/" + c + "-jsonl-repartioned");
			}
			
			JavaRDD<String> input = context.textFile("cikm2020/document-fingerprints-final/" + corpus +"*-jsonl-repartioned");
			
			input.repartition(10000)
				.saveAsTextFile("cikm2020/document-fingerprints-final/" + corpus +"-jsonl.bzip2", BZip2Codec.class);
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/repartition-data");

		return new JavaSparkContext(conf);
	}
}
