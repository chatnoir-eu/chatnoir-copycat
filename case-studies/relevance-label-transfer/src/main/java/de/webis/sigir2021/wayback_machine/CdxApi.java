package de.webis.sigir2021.wayback_machine;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.temporal.ChronoUnit;

import org.apache.commons.io.IOUtils;
import org.apache.commons.lang3.StringUtils;
import org.archive.format.warc.WARCConstants;
import org.archive.format.warc.WARCConstants.WARCRecordType;
import org.archive.io.warc.WARCRecordInfo;
import org.archive.uid.UUIDGenerator;
import org.archive.util.anvl.ANVLRecord;

import lombok.SneakyThrows;
import net.jodah.failsafe.Failsafe;
import net.jodah.failsafe.RetryPolicy;

public class CdxApi {
	public static final RetryPolicy<String> RETRY_POLICY = new RetryPolicy<String>()
			  .handle(Exception.class)
			  .withBackoff(10, 240, ChronoUnit.SECONDS)
			  .withMaxRetries(3);
	
	public static WARCRecordInfo warcRecordForUrlBetween2009And2013(String id, String url) {
		try {
			return warcRecordForUrlBetween2009And2013ThrowsException(id, url);
		} catch(Exception e) {
			return null;
		}
	}
	
	private static WARCRecordInfo warcRecordForUrlBetween2009And2013ThrowsException(String id, String url) {
		String content = snapshotsForUrlBetween2009And2013(url);
		
		WARCRecordInfo ret = new WARCRecordInfo();
		ret.setType(WARCRecordType.response);
		ret.setRecordId(new UUIDGenerator().getRecordID());
		ANVLRecord extraHeaders = new ANVLRecord();
		extraHeaders.addLabelValue(WARCConstants.HEADER_KEY_URI, cdxRequest(url, "2009", "2013").toString());
		extraHeaders.addLabelValue("trec_target_id", id);
		extraHeaders.addLabelValue("trec_target_uri", url);
		ret.setExtraHeaders(extraHeaders);
		InputStream contentStream = new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8));
		ret.setContentLength(content.getBytes(StandardCharsets.UTF_8).length);
		ret.setMimetype("text/plain;charset=UTF-8");
		
		ret.setContentStream(contentStream);
		
		return ret;
	}
	
	public static String snapshotsForUrlBetween2009And2013(String url) {
		URL requestURL = cdxRequest(url, "2009", "2013");

		return snapshotsForUrlFailsave(requestURL);
	}

	private static String snapshotsForUrlFailsave(URL waybackUrl) {
		return snapshotsForUrlFailsave(waybackUrl, RETRY_POLICY);
	}
	
	private static String snapshotsForUrlFailsave(URL waybackUrl, RetryPolicy<String> retryPolicy) {
		return Failsafe.with(retryPolicy).get(() -> snapshotsForUrl(waybackUrl));
	}
	
	@SneakyThrows
	private static String snapshotsForUrl(URL waybackUrl) {
		return IOUtils.toString(waybackUrl, StandardCharsets.UTF_8);
	}
	
	@SneakyThrows
	private static URL cdxRequest(String url, String from, String to) {
		if(url.startsWith("https://")) {
			url = StringUtils.replaceFirst(url, "https://", "");
		}
		if(url.startsWith("http://")) {
			url = StringUtils.replaceFirst(url, "http://", "");
		}
		
		url = URLEncoder.encode(url, StandardCharsets.UTF_8.toString());
		
		return new URL("http://web.archive.org/cdx/search/cdx?url=" + url + "&from=" + from + "&to=" + to);
	}
}
