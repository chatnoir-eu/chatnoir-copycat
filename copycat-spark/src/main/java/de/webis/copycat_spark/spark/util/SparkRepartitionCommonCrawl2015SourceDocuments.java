package de.webis.copycat_spark.spark.util;

import org.apache.hadoop.io.compress.BZip2Codec;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

public class SparkRepartitionCommonCrawl2015SourceDocuments {
	
	private static final String[] CORPORA = new String[] {
			"cc-2015-11-part-0", "cc-2015-11-part-1", "cc-2015-11-part-2",
			"cc-2015-11-part-3", "cc-2015-11-part-4", "cc-2015-11-part-5",
			"cc-2015-11-part-6", "cc-2015-11-part-7", "cc-2015-11-part-8",
			"cc-2015-11-part-9"
	};
	
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			String corpus = "cc-2015-11";
			
			for(String c: CORPORA) {
				context.textFile("cikm2020/document-fingerprints-final/" + c + "-jsonl")
					.repartition(100)
					.saveAsTextFile("cikm2020/document-fingerprints-final/" + c + "-jsonl-repartioned");
			}
			
			JavaRDD<String> input = context.textFile("cikm2020/document-fingerprints-final/" + corpus +"*-jsonl-repartioned");
			
			input.repartition(10000)
				.saveAsTextFile("cikm2020/document-fingerprints-final/" + corpus +"-jsonl.bzip2", BZip2Codec.class);
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/repartition-data");

		return new JavaSparkContext(conf);
	}
}
