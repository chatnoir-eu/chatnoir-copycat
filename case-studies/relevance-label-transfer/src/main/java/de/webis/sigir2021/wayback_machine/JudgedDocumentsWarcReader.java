package de.webis.sigir2021.wayback_machine;

import java.io.ByteArrayOutputStream;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.TimeZone;
import java.util.stream.Collectors;

import org.apache.commons.io.IOUtils;
import org.archive.io.ArchiveRecord;
import org.archive.io.warc.WARCReader;
import org.archive.io.warc.WARCReaderFactory;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import lombok.SneakyThrows;
import net.jodah.failsafe.Failsafe;

public class JudgedDocumentsWarcReader {

	private final Path dir;
	
	public JudgedDocumentsWarcReader(String dir) {
		this.dir = Paths.get(dir);
	}

	@SneakyThrows
	public String entries(int topic, String documentId) {
		Path warcFile = uniqueWARCFile(topic);
		WARCReader reader = WARCReaderFactory.get(warcFile.toFile());
		
		return recordWithId(reader, documentId);
	}

	private String recordWithId(WARCReader reader, String documentId) {
		for(ArchiveRecord record: reader) {
			if(documentId.equalsIgnoreCase((String) record.getHeader().getHeaderFields().get("trec_target_id"))) {
				return content(record);
			}
		}
		
		return null;
	}
	
	@SneakyThrows
	private static String content(ArchiveRecord record) {
		ByteArrayOutputStream ret = new ByteArrayOutputStream();
		record.dump(ret);
		
		return ret.toString(StandardCharsets.UTF_8.toString());
	}

	private Path uniqueWARCFile(int topic) {
		Path parentDir = dir.resolve("topic-" + topic);
		String[] entries = parentDir.toFile().list();
		
		if(entries == null || entries.length != 1) {
			throw new RuntimeException("Invalid dir: " + parentDir);
		}
		
		return parentDir.resolve(entries[0]);
	}

	@SneakyThrows
	public static Date parseDate(String line) {
		try {
			String ret = line.split("\\s+")[1];
			
			return sdf().parse(ret);
		} catch(Exception e) {
			throw new RuntimeException("'" + line + "'");
		}
	}
	
	public static String waybackUrl(String cdxRecord) {
		String date = cdxRecord.split("\\s+")[1];
		
		return "http://web.archive.org/web/" + date + "/" + targetUrl(cdxRecord);
	}
	
	public static String targetUrl(String cdxRecord) {
		return cdxRecord.split("\\s+")[2];
	}
	
	@SneakyThrows
	public static String resolveRedirect(String cdxRecord) {
		if(cdxRecord == null || parseResponseCode(cdxRecord) < 300 || parseResponseCode(cdxRecord)> 400) {
			return null;
		}
		
		return resolveRedirect(new URL(waybackUrl(cdxRecord)), 0);
	}
	
	@SneakyThrows
	public static String resolveRedirect(URL url, int depth) {
		if(depth > 4) {
			return url == null ? null : url.toString();
		}
		
		String html = htmlFailsafe(url.toString());
		String ret = extractRedirectMessageFromHtml(html);
		
		if (ret == null) {
			return url == null ? null : url.toString();
		} else {
			return resolveRedirect(new URL(ret), depth +1);
		}
	}
	
	@SneakyThrows
	private static String htmlFailsafe(String urlStr) {
		URL url = new URL(urlStr);
		return Failsafe.with(CdxApi.RETRY_POLICY).get(() -> html(url));
	}
	
	@SneakyThrows
	private static String html(URL url) {
		return IOUtils.toString(url);
	}

	private static SimpleDateFormat sdf() {
		SimpleDateFormat ret = new SimpleDateFormat("yyyyMMddHHmmss");
		ret.setTimeZone(TimeZone.getTimeZone("Europe/Berlin"));
		
		return ret;
	}
	
	private static Date centerOfCw12Crawl() {
		return parseDate("a 20120401000000 a");
	}

	public static int parseResponseCode(String line) {
		try {
			return Integer.parseInt(line.split("\\s+")[4]);
		} catch(Exception e) {
			return 0;
		}
	}

	public static List<String> keepResponsesInCw12CrawlingTime(List<String> entries) {
		return entries.stream()
				.filter(i -> responseInCw12CrawlingTime(i))
				.collect(Collectors.toList());
	}

	public static boolean responseInCw12CrawlingTime(Date date) {
		return parseDate("a 20111100000000 b").getTime() <= date.getTime() && 
				parseDate("a 20120800000000 b").getTime() >= date.getTime();
	}
	
	private static boolean responseInCw12CrawlingTime(String l) {
		if(l == null || l.trim().isEmpty()) {
			return false;
		}
		
		return responseInCw12CrawlingTime(parseDate(l));
	}

	private static Long diffToCW12(String line) {
		return Math.abs(centerOfCw12Crawl().getTime() - parseDate(line).getTime());
	}

	public String bestMatchingCW12SnapshotOrNull(int topic, String documentId) {
		return bestMatchingCW12SnapshotOrNull(entries(topic, documentId));
	}
	
	public String bestMatchingCW12SnapshotOrNull(String entryString) {
		if(entryString == null) {
			return null;
		}
		
		List<String> entries = Arrays.asList(entryString.split("\n"));
		if(entries.size() == 0) {
			return null;
		}
		
		entries = keepResponsesInCw12CrawlingTime(entries);
		entries = new ArrayList<>(entries.stream().filter(i -> 200 == parseResponseCode(i)).collect(Collectors.toList()));
		if(entries.size() == 0) {
			return null;
		}
		
		Collections.sort(entries, (a,b) -> diffToCW12(a).compareTo(diffToCW12(b)));
		return entries.get(0);
	}

	public String bestMatchingCW12RedirectOrNull(int topic, String documentId) {
		String entryString = entries(topic, documentId);
		if(entryString == null || bestMatchingCW12SnapshotOrNull(entryString) != null) {
			return null;
		}		

		List<String> entries = Arrays.asList(entryString.split("\n"));
		if(entries.size() == 0) {
			return null;
		}
		
		entries = keepResponsesInCw12CrawlingTime(entries);
		entries = new ArrayList<>(entries.stream().filter(i -> parseResponseCode(i) >= 300 && parseResponseCode(i) < 400).collect(Collectors.toList()));
		if(entries.size() == 0) {
			return null;
		}
		
		Collections.sort(entries, (a,b) -> diffToCW12(a).compareTo(diffToCW12(b)));
		return entries.get(0);
	}

	public static String extractRedirectMessageFromHtml(String html) {
		if(html == null) {
			return null;
		}
		
		Document doc = Jsoup.parse(html);
		Elements anchors = doc.select("p[class=impatient]>a[href]");
		
		if(anchors == null || anchors.size() != 1) {
			return null;
		}
		
		return anchors.get(0).attr("href");
	}
}
