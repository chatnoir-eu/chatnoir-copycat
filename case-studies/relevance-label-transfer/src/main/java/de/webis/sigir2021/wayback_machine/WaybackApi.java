package de.webis.sigir2021.wayback_machine;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.net.URL;
import java.net.URLConnection;
import java.time.temporal.ChronoUnit;
import java.util.LinkedHashMap;
import java.util.Map;

import org.apache.commons.io.IOUtils;
import org.apache.commons.lang3.tuple.Pair;
import org.archive.format.warc.WARCConstants;
import org.archive.format.warc.WARCConstants.WARCRecordType;
import org.archive.io.warc.WARCRecordInfo;
import org.archive.uid.UUIDGenerator;
import org.archive.util.anvl.ANVLRecord;

import lombok.SneakyThrows;
import net.jodah.failsafe.Failsafe;
import net.jodah.failsafe.RetryPolicy;

public class WaybackApi {

	public static final RetryPolicy<Pair<ByteArrayOutputStream, Map<String, String>>> RETRY_FINAL = new RetryPolicy<Pair<ByteArrayOutputStream, Map<String, String>>>()
			.handle(Exception.class)
			.withBackoff(3, 100, ChronoUnit.SECONDS).withMaxRetries(4);
	
	
	@SneakyThrows
	public static WARCRecordInfo warcRecord(String id, String cdxLine) {
		Pair<ByteArrayOutputStream, Map<String, String>> response = contentFailsave(cdxLine);
		Map<String, String> headers = response.getRight();
		WARCRecordInfo ret = new WARCRecordInfo();
		ret.setType(WARCRecordType.response);
		ret.setRecordId(new UUIDGenerator().getRecordID());
		InputStream contentStream = new ByteArrayInputStream(response.getLeft().toByteArray());
		
		ret.setContentLength(Integer.parseInt(headers.get("Content-Length")));
		ret.setMimetype(headers.get("Content-Type"));
		
		ANVLRecord extraHeaders = new ANVLRecord();
		extraHeaders.addLabelValue(WARCConstants.HEADER_KEY_URI, JudgedDocumentsWarcReader.waybackUrl(cdxLine));
		extraHeaders.addLabelValue("trec_target_id", id);
		extraHeaders.addLabelValue("trec_target_uri", JudgedDocumentsWarcReader.targetUrl(cdxLine));
		for(String header: headers.keySet()) {
			if(header == null || header.equalsIgnoreCase("Content-Length") || header.equalsIgnoreCase("Content-Type")) {
				continue;
			}
			
			extraHeaders.addLabelValue(header, headers.get(header));
		}
		
		ret.setExtraHeaders(extraHeaders);
		
		ret.setContentStream(contentStream);
		
		return ret;
	}
	
	@SneakyThrows
	public static WARCRecordInfo warcRecord(String id, URL url) {
		Pair<ByteArrayOutputStream, Map<String, String>> response = contentFailsave(url);
		Map<String, String> headers = response.getRight();
		WARCRecordInfo ret = new WARCRecordInfo();
		ret.setType(WARCRecordType.response);
		ret.setRecordId(new UUIDGenerator().getRecordID());
		InputStream contentStream = new ByteArrayInputStream(response.getLeft().toByteArray());
		
		ret.setContentLength(Integer.parseInt(headers.get("Content-Length")));
		ret.setMimetype(headers.get("Content-Type"));
		
		ANVLRecord extraHeaders = new ANVLRecord();
		extraHeaders.addLabelValue(WARCConstants.HEADER_KEY_URI, url.toString());
		extraHeaders.addLabelValue("trec_target_id", id);
		extraHeaders.addLabelValue("trec_target_uri", url.toString());
		for(String header: headers.keySet()) {
			if(header == null || header.equalsIgnoreCase("Content-Length") || header.equalsIgnoreCase("Content-Type")) {
				continue;
			}
			
			extraHeaders.addLabelValue(header, headers.get(header));
		}
		
		ret.setExtraHeaders(extraHeaders);
		
		ret.setContentStream(contentStream);
		
		return ret;
	}
	
	@SneakyThrows
	private static Pair<ByteArrayOutputStream, Map<String, String>> contentFailsave(String cdxLine) {
		URL waybackUrl = new URL(JudgedDocumentsWarcReader.waybackUrl(cdxLine));
		return contentFailsave(waybackUrl);
	}
	
	private static Pair<ByteArrayOutputStream, Map<String, String>> contentFailsave(URL waybackUrl) {
		return Failsafe.with(RETRY_FINAL).get(() -> content(waybackUrl));
	}
	
	@SneakyThrows
	private static Pair<ByteArrayOutputStream, Map<String, String>> content(URL waybackUrl) {
		
		URLConnection conn = waybackUrl.openConnection();
		
		ByteArrayOutputStream ret = new ByteArrayOutputStream();
		IOUtils.copy(conn.getInputStream(), ret);
		
		Map<String, String> headers = new LinkedHashMap<>();
		for(String header: conn.getHeaderFields().keySet()) {
			headers.put(header, conn.getHeaderField(header));
		}
		
		return Pair.of(ret, headers);
	}
}
