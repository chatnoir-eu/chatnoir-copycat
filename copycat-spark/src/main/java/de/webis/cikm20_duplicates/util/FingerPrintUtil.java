package de.webis.cikm20_duplicates.util;

import java.io.Serializable;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import de.aitools.ir.fingerprinting.hashfunction.SimHash;
import de.aitools.ir.fingerprinting.representation.HashVector;
import de.aitools.ir.fingerprinting.representation.HashVectorSha3;
import de.aitools.ir.fingerprinting.representer.Hash;
import de.webis.cikm20_duplicates.spark.eval.SparkEvaluateSimHashFeatures;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import de.webis.trec_ndd.util.NGramms;
import info.debatty.java.lsh.MinHash;
import lombok.experimental.UtilityClass;

/**
 * 
 * @author Maik Fr√∂be
 *
 */
@UtilityClass
public class FingerPrintUtil {
	public static interface Fingerprinter<T extends Comparable<T>> extends Serializable {
		public List<T> fingerprint(CollectionDocument doc);
		public List<T> fingerprint(List<String> features);
		
		public double similarity(List<T> a, List<T> b);
		public String fingerprinterName();
		
		public default double similarity(CollectionDocument a, CollectionDocument b) {
			List<T> fingerprintA = fingerprint(a);
			List<T> fingerprintB = fingerprint(b);
			
			return similarity(fingerprintA, fingerprintB);
		}
	}

	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> minHashFingerPrinting(long seed) {
		int dict_size = 1;
		int size = 12;
		MinHash minHash = new MinHash(size, dict_size, seed);
		
		return new Fingerprinter<Integer>() {
			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				return IntStream.of(minHash.signature(docToElementSet(doc))).boxed().collect(Collectors.toList());
			}
			
			@Override
			public List<Integer> fingerprint(List<String> features) {
				throw new RuntimeException();
			}

			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				int[] aArray = a.stream().mapToInt(i -> i).toArray();
				int[] bArray = b.stream().mapToInt(i -> i).toArray();
				
				return minHash.similarity(aArray, bArray);
			}
			
			public String fingerprinterName() {
				return "MinHashWithJavaHash";
			}
		};
	}
	
	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> simHashFingerPrinting(int bitsInSimHash, int k) {
		return new Fingerprinter<Integer>() {

			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				byte[] hash = hashVector(doc, bitsInSimHash);
				
				return HashTransformationUtil.hashToIntegers(hash, k);
			}

			@Override
			public List<Integer> fingerprint(List<String> features) {
				byte[] hash = hashVector(features, bitsInSimHash);
				
				return HashTransformationUtil.hashToIntegers(hash, k);
			}
			
			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				byte[] aArr = HashTransformationUtil.integersToHash(a);
				byte[] bArr = HashTransformationUtil.integersToHash(b);
				
				return 1.0 - (((double)Hash.getHammingDistance(aArr, bArr))/((double)aArr.length*8));
			}
			
			public String fingerprinterName() {
				return "64BitK3SimHashOneGramms";
			}
		};
	}
	
	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> productionFingerpringint(int bitsInSimHash, int k) {
		return new Fingerprinter<Integer>() {

			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				byte[] hash = hashVector(SparkEvaluateSimHashFeatures.theeAndFiveGramms(doc), bitsInSimHash);
				
				return HashTransformationUtil.hashToIntegers(hash, k);
			}

			@Override
			public List<Integer> fingerprint(List<String> features) {
				byte[] hash = hashVector(features, bitsInSimHash);
				
				return HashTransformationUtil.hashToIntegers(hash, k);
			}
			
			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				byte[] aArr = HashTransformationUtil.integersToHash(a);
				byte[] bArr = HashTransformationUtil.integersToHash(b);
				
				return 1.0 - (((double)Hash.getHammingDistance(aArr, bArr))/((double)aArr.length*8));
			}
			
			public String fingerprinterName() {
				return "64BitK3SimHashThreeAndFiveGramms";
			}
		};
	}
	
	static byte[] hashVector(CollectionDocument doc, int bitsInSimHash) {
		return new SimHash().hash(internalHashVector(doc, bitsInSimHash)).getArray();
	}
	
	static byte[] hashVector(List<String> terms, int bitsInSimHash) {
		return new SimHash().hash(internalHashVector(terms, bitsInSimHash)).getArray();
	}
	
	public static HashVector internalHashVector(CollectionDocument doc, int bitsInSimHash) {
		List<String> terms = NGramms.tokenize(doc.getFullyCanonicalizedContent());
		return internalHashVector(terms, bitsInSimHash);
	}
	
	public static HashVector internalHashVector(List<String> terms, int bitsInSimHash) {
		return HashVectorSha3.toVector(terms, bitsInSimHash);
	}
	
	private static Set<Integer> docToElementSet(CollectionDocument doc) {
		return NGramms.build8Gramms(doc.getFullyCanonicalizedContent()).stream()
				.map(i -> i.getMd5Hash().hashCode())
				.collect(Collectors.toSet());
	}
}
