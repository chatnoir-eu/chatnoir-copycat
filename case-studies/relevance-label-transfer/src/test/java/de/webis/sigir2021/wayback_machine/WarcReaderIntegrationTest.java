package de.webis.sigir2021.wayback_machine;

import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.TimeZone;

import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.Test;

import de.webis.sigir2021.wayback_machine.JudgedDocumentsWarcReader;

public class WarcReaderIntegrationTest {
	@Test
	public void approveSnapshotsForCw19493() {
		String documentId = "clueweb09-en0000-02-19493";
		int topic = 1;
		
		String actual = new JudgedDocumentsWarcReader("src/test/resources/sample-input").entries(topic, documentId);
		Approvals.verify(actual);
	}
	
	@Test
	public void approveSnapshotsForCw02851() {
		String documentId = "clueweb09-enwp03-13-02851";
		int topic = 1;
		
		String actual = new JudgedDocumentsWarcReader("src/test/resources/sample-input").entries(topic, documentId);
		Approvals.verify(actual);
	}
	
	@Test
	public void testThatCorrectDocumentIsSelectedFromMultipleSnapshots() {
		String documentId = "clueweb09-enwp03-13-02851";
		int topic = 1;
		
		String actual = new JudgedDocumentsWarcReader("src/test/resources/sample-input").bestMatchingCW12SnapshotOrNull(topic, documentId);
		Assert.assertEquals("org,wikipedia,en)/wiki/ruth_ndesandjo 20120315144017 http://en.wikipedia.org:80/wiki/Ruth_Ndesandjo text/html 200 6GYMDNUDRFTIPSHSPIYIOPJAGODHSZXA 71883", actual);
	}
	
	@Test
	public void testThatNoRedirectRequestIsExtractedForDocumentWith200erResponse() {
		String documentId = "clueweb09-enwp03-13-02851";
		int topic = 1;
		
		String actual = new JudgedDocumentsWarcReader("src/test/resources/sample-input").bestMatchingCW12RedirectOrNull(topic, documentId);
		Assert.assertNull(actual);
	}
	
	@Test
	public void testThatCorrectRedirectRequestIsExtractedForDocumentWith301erResponse() {
		String documentId = "clueweb09-en0007-79-21751";
		String expected = "com,jamespot)/a/125441-barack-obama-s-family-tree.html 20120113225712 http://www.jamespot.com/a/125441-Barack-Obama-s-Family-Tree.html text/html 301 3I42H3S6NNFQ2MSVX7XZKYAYSCX5QBYJ 484";
		int topic = 1;
		
		String actual = new JudgedDocumentsWarcReader("src/test/resources/sample-input").bestMatchingCW12RedirectOrNull(topic, documentId);
		Assert.assertEquals(expected, actual);
	}
	
	@Test
	public void testThatCorrectDocumentIsSelectedFromMultipleSnapshots2() {
		String documentId = "clueweb09-en0000-02-19493";
		int topic = 1;
			
		String actual = new JudgedDocumentsWarcReader("src/test/resources/sample-input").bestMatchingCW12SnapshotOrNull(topic, documentId);
		Assert.assertNull(actual);
	}
	
	@Test
	public void testParsingOfDate() {
		String expected = "2011/9/910-11:09:49";
		Date actual = JudgedDocumentsWarcReader.parseDate("com,cellphonenews)/category/usa_canada_phones/3g_cell_phones 20110910110949 http://www.cellphonenews.com/category/usa_canada_phones/3g_cell_phones text/html 200 E3OYLZFLBA3DHLJU23RQ3ZGHGDUTGGCK 12439");
		
		Assert.assertEquals(expected, sdf().format(actual));
	}
	
	@Test
	public void testParsingOfDate2() {
		String expected = "2009/1/127-05:16:46";
		Date actual = JudgedDocumentsWarcReader.parseDate("com,mahalo)/abo_obama 20090127051646 http://mahalo.com:80/Abo_Obama text/html 302 POPLDWWERZ2PUX2BRPCFNS2BB6ELQHMY 309");
		
		Assert.assertEquals(expected, sdf().format(actual));
	}
	
	private static SimpleDateFormat sdf() {
		SimpleDateFormat ret = new SimpleDateFormat("yyyy/M/Mdd-HH:mm:ss");
		ret.setTimeZone(TimeZone.getTimeZone("Europe/Berlin"));
		
		return ret;
	}

	@Test
	public void testParsingOfResponseCode() {
		int expected = 200;
		int actual = JudgedDocumentsWarcReader.parseResponseCode("com,cellphonenews)/category/usa_canada_phones/3g_cell_phones 20110910110949 http://www.cellphonenews.com/category/usa_canada_phones/3g_cell_phones text/html 200 E3OYLZFLBA3DHLJU23RQ3ZGHGDUTGGCK 12439");
		
		Assert.assertEquals(expected, actual);
	}
	
	@Test
	public void testParsingOfResponseCodeForNon302Code() {
		int expected = 302;
		int actual = JudgedDocumentsWarcReader.parseResponseCode("com,mahalo)/abo_obama 20090127051646 http://mahalo.com:80/Abo_Obama text/html 302 POPLDWWERZ2PUX2BRPCFNS2BB6ELQHMY 309");
		
		Assert.assertEquals(expected, actual);
	}
	
	@Test
	public void testThatEntriesOutsideOfCw12CrawlingRangeAreRemoved() {
		List<String> entries = Arrays.asList(
			"com,mahalo)/abo_obama 20090127051646 http://mahalo.com:80/Abo_Obama text/html 200 POPLDWWERZ2PUX2BRPCFNS2BB6ELQHMY 309",
			"com,mahalo)/abo_obama 20130127051646 http://mahalo.com:80/Abo_Obama text/html 200 POPLDWWERZ2PUX2BRPCFNS2BB6ELQHMY 309"
		);
		
		List<String> remaining = JudgedDocumentsWarcReader.keepResponsesInCw12CrawlingTime(entries);
		
		Assert.assertTrue(remaining.isEmpty());
	}
	
	@Test
	public void testThatEntriesWithinCw12CrawlingRangeAreKept() {
		List<String> entries = Arrays.asList(
			"com,mahalo)/abo_obama 20111027051646 http://mahalo.com:80/Abo_Obama text/html 200 POPLDWWERZ2PUX2BRPCFNS2BB6ELQHMY 309",
			"com,mahalo)/abo_obama 20111227051646 http://mahalo.com:80/Abo_Obama text/html 200 POPLDWWERZ2PUX2BRPCFNS2BB6ELQHMY 309",
			"com,mahalo)/abo_obama 20120727051646 http://mahalo.com:80/Abo_Obama text/html 200 POPLDWWERZ2PUX2BRPCFNS2BB6ELQHMY 309",
			"com,mahalo)/abo_obama 20121227051646 http://mahalo.com:80/Abo_Obama text/html 200 POPLDWWERZ2PUX2BRPCFNS2BB6ELQHMY 309"
		);
		List<String> expected = Arrays.asList(
			"com,mahalo)/abo_obama 20111227051646 http://mahalo.com:80/Abo_Obama text/html 200 POPLDWWERZ2PUX2BRPCFNS2BB6ELQHMY 309",
			"com,mahalo)/abo_obama 20120727051646 http://mahalo.com:80/Abo_Obama text/html 200 POPLDWWERZ2PUX2BRPCFNS2BB6ELQHMY 309"
		);
		
		List<String> actual = JudgedDocumentsWarcReader.keepResponsesInCw12CrawlingTime(entries);
		
		Assert.assertEquals(expected, actual);
	}
}
