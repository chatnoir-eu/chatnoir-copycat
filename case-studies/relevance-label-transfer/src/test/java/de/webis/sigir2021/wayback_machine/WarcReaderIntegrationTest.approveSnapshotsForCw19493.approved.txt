com,tnr,blogs)/tnr/blogs/the_plank/archive/2009/01/07/did-bush-snub-the-obama-family.aspx 20111130215150 http://blogs.tnr.com/tnr/blogs/the_plank/archive/2009/01/07/did-bush-snub-the-obama-family.aspx text/html 302 IXNINMU5FLAXAS4OXR5POJHCJWJFYRJL 457
com,tnr,blogs)/tnr/blogs/the_plank/archive/2009/01/07/did-bush-snub-the-obama-family.aspx 20111202185203 http://blogs.tnr.com/tnr/blogs/the_plank/archive/2009/01/07/did-bush-snub-the-obama-family.aspx text/html 302 IXNINMU5FLAXAS4OXR5POJHCJWJFYRJL 457
com,tnr,blogs)/tnr/blogs/the_plank/archive/2009/01/07/did-bush-snub-the-obama-family.aspx 20111204213220 http://blogs.tnr.com/tnr/blogs/the_plank/archive/2009/01/07/did-bush-snub-the-obama-family.aspx text/html 302 IXNINMU5FLAXAS4OXR5POJHCJWJFYRJL 458
